cpp/src/join/unique_hash_join.cu

```diff
@@ -0,0 +1,334 @@
+unique_hash_join<Hasher, HasNested>::unique_hash_join(cudf::table_view const& build,
+                                                      cudf::table_view const& probe,
+                                                      bool has_nulls,
+                                                      cudf::null_equality compare_nulls,
+                                                      rmm::cuda_stream_view stream)
+  : _has_nulls{has_nulls},
+    _is_empty{build.num_rows() == 0},
+    _nulls_equal{compare_nulls},
+    _build{build},
+    _probe{probe},
+    _preprocessed_build{
+      cudf::experimental::row::equality::preprocessed_table::create(_build, stream)},
+    _preprocessed_probe{
+      cudf::experimental::row::equality::preprocessed_table::create(_probe, stream)},
+    _hash_table{::compute_hash_table_size(build.num_rows()),
+                cuco::empty_key{cuco::pair{std::numeric_limits<hash_value_type>::max(),
+                                           lhs_index_type{JoinNoneValue}}},
+                prepare_device_equal<HasNested>(
+                  _preprocessed_build, _preprocessed_probe, has_nulls, compare_nulls),
+                {},
+                cudf::detail::cuco_allocator{stream},
+                stream.value()}
+{
+  CUDF_FUNC_RANGE();
+  CUDF_EXPECTS(0 != this->_build.num_columns(), "Hash join build table is empty");
+
+  if (this->_is_empty) { return; }
+
+  auto const row_hasher = experimental::row::hash::row_hasher{this->_preprocessed_build};
+  auto const d_hasher   = row_hasher.device_hasher(nullate::DYNAMIC{this->_has_nulls});
+
+  auto const iter = cudf::detail::make_counting_transform_iterator(
+    0, build_keys_fn<decltype(d_hasher), lhs_index_type>{d_hasher});
+
+  size_type const build_table_num_rows{build.num_rows()};
+  if (this->_nulls_equal == cudf::null_equality::EQUAL or (not cudf::nullable(this->_build))) {
+    this->_hash_table.insert_async(iter, iter + build_table_num_rows, stream.value());
+  } else {
+    auto stencil = thrust::counting_iterator<size_type>{0};
+    auto const row_bitmask =
+      cudf::detail::bitmask_and(this->_build, stream, rmm::mr::get_current_device_resource()).first;
+    auto const pred =
+      cudf::detail::row_is_valid{reinterpret_cast<bitmask_type const*>(row_bitmask.data())};
+
+    // insert valid rows
+    this->_hash_table.insert_if_async(
+      iter, iter + build_table_num_rows, stencil, pred, stream.value());
+  }
+}
+
+template <typename Hasher, cudf::has_nested HasNested>
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+unique_hash_join<Hasher, HasNested>::inner_join(std::optional<std::size_t> output_size,
+                                                rmm::cuda_stream_view stream,
+                                                rmm::mr::device_memory_resource* mr) const
+{
+  CUDF_FUNC_RANGE();
+
+  size_type const probe_table_num_rows{this->_probe.num_rows()};
```

Comment:
Should return early here if either build or probe side is empty.  Currently crashes with `cudf/detail/utilities/cuda.cuh:66: num_blocks must be > 0` error if probe side is empty.  Need tests for empty build and empty probe.